src/boot_O2/__osMalloc.c

```diff
@@ -45,181 +49,353 @@ ArenaNode* ArenaImpl_GetLastBlock(Arena* arena) {
     return last;
 }
 
-void __osMallocInit(Arena* arena, void* start, size_t size) {
-    bzero(arena, sizeof(*arena));
+/**
+ * @brief Initializes \p arena using \p heap.
+ *
+ * @param arena  The Arena to initialize.
+ * @param heap   The heap to use.
+ * @param size   The size of the heap.
+ */
+void __osMallocInit(Arena* arena, void* heap, size_t size) {
+    bzero(arena, sizeof(Arena));
+
     ArenaImpl_LockInit(arena);
-    __osMallocAddBlock(arena, start, size);
-    arena->isInit = 1;
+
+    __osMallocAddBlock(arena, heap, size);
+    arena->isInit = true;
 }
 
-void __osMallocAddBlock(Arena* arena, void* start, s32 size) {
-    s32 diff;
-    s32 size2;
+void __osMallocAddBlock(Arena* arena, void* heap, size_t size) {
```

Comment:
This function initializes the first block, "add block" suggests to me that it's used whenever a new block is added, which is not the case. I'm not totally sure on a succinct name however.

Reply:
You are right.
But at the same time, this function _has_ the functionality to add a new block, so the name isn't technically wrong, it just happens that each an allocating function "adds" a block, it does it manually by spliting one of the already existing ones.

This function could theoretically be used to add blocks new blocks to completely unrelated heaps, it just happens that the game never does this.

What about `__osMallocAddHeap`?

Reply:
Also, this name comes from the Animal Forest map.

I took a quick look at AF and I was able to find this function (ROM 0x008060, VRAM: 0x8002CC60 for anybody interested). It is exactly the same (and it matches without -g3. https://decomp.me/scratch/z9aiE).

There's a catch tho. I compared the other functions in AF, and looks like there are a few functions in the AF map missing from the AF rom, which shouldn't happen, right? Also, the VRAMs from the map doesn't match the VRAMs from the ROM itself (which may not be a huge deal? idk).

My conclusion is, what we have been calling `__osMalloc` all this time is not the real `__osMalloc` that the function map refers to. If this is true, I propose to rename all the functions in this file to something more ad hoc to our naming scheme (`SystemName_Func`). What do you think?

Reply:
Perhaps I'm wrong, but I'm pretty sure this cannot add a new block, only the first block, which is an important distinction. It looks to me that if this function were called more than once for a particular Arena, the arena would lose the ability to keep track of prior blocks in the other heap space. If that's correct, I suggest `__osMallocAddFirstBlock`. Otherwise, I think `__osMallocAddHeap` is good.

Regarding the AF map, the map comes from the port to Gamecube, so there are bound to be a number of differences when looking at the N64 version. It's likely AF on the Gamecube uses a modified `__osMalloc` but it should certainly have it's roots in the N64 version.
I see that `__osMallocAddBlock` is the original name as you pointed out, which is surely why it was named that way in the first place. I don't believe we should stick to it though if it's not representative of the purpose of the function, which I don't think it really is. Perhaps it would be nice however to comment it like
`// Original name: __osMallocAddBlock`

Reply:
Ah I see what you mean now, I failed to notice that it does in fact link with existing blocks via `ArenaImpl_GetLastBlock`. In that case, I think `__osMallocAddHeap` is suitable.

Reply:
Sounds good. I'll rename it to `__osMallocAddHeap`